In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os, sys
import datetime

In [2]:
df = pd.read_csv(r'G:/project/data_03052019/Computers - Software/TCS.csv', header = None)

In [3]:
df.head()

0       1       2       3       4          5    6    7   8   9
0  25 Aug 2004  149.84  149.84  122.38  123.50  136930976  NaN  NaN NaN NaN
1  26 Aug 2004  124.00  124.63  121.92  122.38   40443200  NaN  NaN NaN NaN
2  27 Aug 2004  122.80  122.80  119.82  120.34   30646000  NaN  NaN NaN NaN
3  30 Aug 2004  121.24  123.75  120.63  123.35   24465208  NaN  NaN NaN NaN
4  31 Aug 2004  123.32  123.75  122.00  123.52   21194656  NaN  NaN NaN NaN

In [4]:
df.columns = ['date', 'open', 'high', 'low', 'close', 'volume', 'Bonus', 'dividend', 'Rights : Ratio', 'split' ]

In [5]:
df.sample(10)

date     open     high      low    close   volume Bonus dividend  \
863   01 Feb 2008   221.25   235.28   218.33   232.38  4169520   NaN      NaN   
1191  15 Jun 2009   191.18   196.25   188.75   194.50  6595868   NaN      NaN   
1669  13 May 2011   556.00   571.15   556.00   563.55  2365488   NaN      NaN   
2782  05 Nov 2015  1265.40  1267.48  1240.00  1241.33  1596098   NaN      NaN   
2632  30 Mar 2015  1262.50  1284.95  1260.00  1282.08  1339006   NaN      NaN   
586   22 Dec 2006   283.75   290.00   282.03   288.49  1750016   NaN      NaN   
1070  08 Dec 2008   133.75   136.57   128.10   130.57  7621404   NaN      NaN   
994   13 Aug 2008   200.00   209.43   200.00   207.08  3697968   NaN      NaN   
3070  05 Jan 2017  1192.00  1194.98  1160.20  1167.20  2216404   NaN      NaN   
2983  30 Aug 2016  1255.00  1274.90  1250.63  1272.08  1444026   NaN      NaN   

      Rights : Ratio  split  
863              NaN    NaN  
1191             NaN    NaN  
1669             NaN    NaN  
2782             NaN    NaN  
2632             NaN    NaN  
586              NaN    NaN  
1070             NaN    NaN  
994              NaN    NaN  
3070             NaN    NaN  
2983             NaN    NaN

In [6]:
df.rename({'Bonus' : 'bonus', 'Rights : Ratio' : 'rights'}, axis=1, inplace=True)

In [7]:
df.sample(5)

date     open     high      low    close   volume bonus dividend  \
1436  10 Jun 2010   365.00   376.50   365.00   375.25  1796036   NaN      NaN   
1451  01 Jul 2010   375.00   375.00   362.88   366.00  3407744   NaN      NaN   
341   02 Jan 2006   213.50   213.87   210.17   210.95  1653504   NaN      NaN   
2387  26 Mar 2014  1074.03  1079.25  1043.78  1046.75  3192350   NaN      NaN   
2422  21 May 2014  1030.60  1044.50  1027.90  1041.43  4116946   NaN      NaN   

      rights  split  
1436     NaN    NaN  
1451     NaN    NaN  
341      NaN    NaN  
2387     NaN    NaN  
2422     NaN    NaN

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3644 entries, 0 to 3643
Data columns (total 10 columns):
date        3644 non-null object
open        3644 non-null float64
high        3644 non-null float64
low         3644 non-null float64
close       3644 non-null float64
volume      3644 non-null int64
bonus       3 non-null object
dividend    58 non-null object
rights      0 non-null float64
split       0 non-null float64
dtypes: float64(6), int64(1), object(3)
memory usage: 284.8+ KB


In [9]:
df.describe()

open         high          low        close        volume  \
count  3644.000000  3644.000000  3644.000000  3644.000000  3.644000e+03   
mean    746.440686   754.816216   737.446984   746.127302  3.778019e+06   
std     541.002422   545.703824   535.885433   540.920118  4.093183e+06   
min     112.000000   116.120000   103.840000   111.550000  5.603400e+04   
25%     252.250000   254.997500   247.780000   250.800000  1.923974e+06   
50%     586.625000   593.205000   579.190000   586.940000  2.858334e+06   
75%    1222.062500  1234.025000  1211.500000  1220.737500  4.500502e+06   
max    2255.000000  2275.950000  2230.100000  2260.350000  1.369310e+08   

       rights  split  
count     0.0    0.0  
mean      NaN    NaN  
std       NaN    NaN  
min       NaN    NaN  
25%       NaN    NaN  
50%       NaN    NaN  
75%       NaN    NaN  
max       NaN    NaN

In [10]:
df['date']= pd.to_datetime(df['date'])

In [11]:
df.head(3)

date    open    high     low   close     volume bonus dividend  \
0 2004-08-25  149.84  149.84  122.38  123.50  136930976   NaN      NaN   
1 2004-08-26  124.00  124.63  121.92  122.38   40443200   NaN      NaN   
2 2004-08-27  122.80  122.80  119.82  120.34   30646000   NaN      NaN   

   rights  split  
0     NaN    NaN  
1     NaN    NaN  
2     NaN    NaN

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3644 entries, 0 to 3643
Data columns (total 10 columns):
date        3644 non-null datetime64[ns]
open        3644 non-null float64
high        3644 non-null float64
low         3644 non-null float64
close       3644 non-null float64
volume      3644 non-null int64
bonus       3 non-null object
dividend    58 non-null object
rights      0 non-null float64
split       0 non-null float64
dtypes: datetime64[ns](1), float64(6), int64(1), object(2)
memory usage: 284.8+ KB


In [13]:
df[df['bonus'].notnull()]

date     open     high     low    close   volume      bonus  \
484  2006-07-28   235.03   243.74   231.0   232.12  3019608  Ratio 1:1   
1192 2009-06-16   191.65   196.75   185.6   194.58  4344006  Ratio 1:1   
3417 2018-05-31  1734.00  1759.05  1726.1  1741.05  5049371  Ratio 1:1   

          dividend  rights  split  
484   Interim 300%     NaN    NaN  
1192    Final 500%     NaN    NaN  
3417   Final 2900%     NaN    NaN

In [14]:
df['bonus'] = df['bonus'].astype('category')

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3644 entries, 0 to 3643
Data columns (total 10 columns):
date        3644 non-null datetime64[ns]
open        3644 non-null float64
high        3644 non-null float64
low         3644 non-null float64
close       3644 non-null float64
volume      3644 non-null int64
bonus       3 non-null category
dividend    58 non-null object
rights      0 non-null float64
split       0 non-null float64
dtypes: category(1), datetime64[ns](1), float64(6), int64(1), object(1)
memory usage: 259.9+ KB


In [16]:
df[df['dividend'].notnull()].sample(5)

date     open     high      low    close   volume bonus  \
3450 2018-07-17  1989.00  2009.00  1981.00  2002.15  3315492   NaN   
791  2007-10-22   279.25   279.25   266.28   267.99  3836808   NaN   
1845 2012-01-25   547.45   554.98   545.53   549.00  5845238   NaN   
2434 2014-06-06  1050.08  1053.90  1028.00  1042.13  3361586   NaN   
2037 2012-10-31   654.55   658.40   652.00   657.75   999536   NaN   

          dividend  rights  split  
3450  Interim 400%     NaN    NaN  
791   Interim 300%     NaN    NaN  
1845  Interim 300%     NaN    NaN  
2434   Final 2000%     NaN    NaN  
2037  Interim 300%     NaN    NaN

In [17]:
df[['split', 'rights']] = df[['split', 'rights']].astype('category')

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3644 entries, 0 to 3643
Data columns (total 10 columns):
date        3644 non-null datetime64[ns]
open        3644 non-null float64
high        3644 non-null float64
low         3644 non-null float64
close       3644 non-null float64
volume      3644 non-null int64
bonus       3 non-null category
dividend    58 non-null object
rights      0 non-null category
split       0 non-null category
dtypes: category(3), datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 210.3+ KB


In [19]:
df['date'].min()

Timestamp('2004-08-25 00:00:00')

TCS shares got listed with NSE on 25th Aug 2004

In [20]:
df[['open', 'close', 'high']].max()

open     2255.00
close    2260.35
high     2275.95
dtype: float64

In [21]:
df.iloc[df['high'].idxmax()][['date', 'high']]

date    2018-10-01 00:00:00
high                2275.95
Name: 3500, dtype: object

TCS had it's all time high on 01st oct 2018 and the value is 2275.95

In [22]:
df.iloc[df['low'].idxmin()][['date', 'low']]

date    2008-10-27 00:00:00
low                  103.84
Name: 1043, dtype: object

TCS had it's all time low on 27th oct 2008 and the value is 103.84

In [23]:
df[df['volume'] == df['volume'].max()]

date    open    high     low  close     volume bonus dividend rights  \
0 2004-08-25  149.84  149.84  122.38  123.5  136930976   NaN      NaN    NaN   

  split  
0   NaN

Highest volume traded on TCS was 136930976 and is on it's first day of releasing IPO

In [24]:
df.iloc[(( df['close'].shift(1) - df['close'] ) / df['close'] * 100).idxmax()]

date        2008-04-22 00:00:00
open                     246.24
high                     246.24
low                       220.5
close                     221.8
volume                 13785196
bonus                       NaN
dividend                    NaN
rights                      NaN
split                       NaN
Name: 915, dtype: object

In [25]:
((df['close'].shift(1) - df['close'] ) / df['close'] * 100).max()

11.970243462578892

In [26]:
((df['close'].shift(1) - df['close'] ) / df['close'] * 100).min()

-13.41614238855729

In [27]:
df.iloc[(( df['close'].shift(1) - df['close'] ) / df['close'] * 100).idxmin()]

date        2009-05-18 00:00:00
open                     191.65
high                      197.9
low                      173.75
close                    185.97
volume                    73380
bonus                       NaN
dividend                    NaN
rights                      NaN
split                       NaN
Name: 1171, dtype: object

Till date, stock was most volatile on 18th Mar, 2009 which had a negative movement of 13.4% and maximum positive movement experienced by a stock in a single day is nearly 12%.

In [28]:
(df['date'] - df['date'].shift(1)).max()

Timedelta('6 days 00:00:00')

Maximum consecutive days on which the market was shut is 6 days

In [29]:
import math
import datetime

def func(x):
    mon = int(datetime.datetime.strftime( x, '%m'))
    return math.ceil(mon / 3)

In [30]:
df['quarter'] =  'Q' + df['date'].apply(func).apply(str)

In [31]:
df.sample(4)

date    open    high     low   close   volume bonus dividend  \
743  2007-08-13  287.50  287.50  281.29  283.63  1205524   NaN      NaN   
1898 2012-04-13  545.68  552.35  530.63  534.18  8994680   NaN      NaN   
778  2007-10-03  262.50  273.75  262.48  269.80  5122872   NaN      NaN   
1696 2011-06-21  537.50  555.35  531.60  553.38  3425946   NaN      NaN   

     rights split quarter  
743     NaN   NaN      Q3  
1898    NaN   NaN      Q2  
778     NaN   NaN      Q4  
1696    NaN   NaN      Q2

In [32]:
vals_to_replace = {'Q2':'Q1', 'Q3':'Q2', 'Q4':'Q3', 'Q1':'Q4'}
df['quarter'] = df['quarter'].map(vals_to_replace)

In [33]:
df.sample(5)

date    open     high      low    close   volume bonus dividend  \
3088 2017-02-01  1110.0  1119.98  1077.78  1083.95  4864152   NaN      NaN   
10   2004-09-08   124.5   125.10   123.89   124.22  6593984   NaN      NaN   
3351 2018-02-22  1526.0  1539.50  1511.25  1525.10  3463010   NaN      NaN   
2310 2013-12-05  1008.0  1016.78   991.53   993.58  1397190   NaN      NaN   
1549 2010-11-18   536.0   536.00   520.00   521.40  2765160   NaN      NaN   

     rights split quarter  
3088    NaN   NaN      Q4  
10      NaN   NaN      Q2  
3351    NaN   NaN      Q4  
2310    NaN   NaN      Q3  
1549    NaN   NaN      Q3

In [34]:
df.iloc[(df['volume'] * df['close']).idxmax()]

date        2018-03-13 00:00:00
open                    1462.45
high                    1469.85
low                        1436
close                    1443.4
volume                 88067154
bonus                       NaN
dividend                    NaN
rights                      NaN
split                       NaN
quarter                      Q4
Name: 3363, dtype: object

In [35]:
(df['volume'] * df['close']).sort_values(ascending = False).head(5)

3363    1.271161e+11
3446    4.534724e+10
3389    3.141434e+10
3390    2.588180e+10
3393    2.147452e+10
dtype: float64

Most active in terms of funds turnover was on 13th Mar 18 and registered nearly 12711Cr rupees.

In [36]:
df['quarter'] = df['quarter'].astype('category')

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3644 entries, 0 to 3643
Data columns (total 11 columns):
date        3644 non-null datetime64[ns]
open        3644 non-null float64
high        3644 non-null float64
low         3644 non-null float64
close       3644 non-null float64
volume      3644 non-null int64
bonus       3 non-null category
dividend    58 non-null object
rights      0 non-null category
split       0 non-null category
quarter     3644 non-null category
dtypes: category(4), datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 214.0+ KB


In [38]:
df.groupby('quarter')['volume'].mean()

quarter
Q1    4.002319e+06
Q2    3.901983e+06
Q3    3.621290e+06
Q4    3.594913e+06
Name: volume, dtype: float64

In [39]:
df['close'].tail(1).iloc[0] /  df['close'].head(1).iloc[0] * 100

1726.3157894736842

The price of the stock has gone up by 17.4 times(1740%). i.e if we would have invested 1 lakh on the first day of IPO then, current worth of our stocks would be roughly around 17.4 lakhs

In [40]:
from dateutil.relativedelta import relativedelta

no_of_days = 365
start_date = datetime.datetime.strptime('01 Apr 2005', '%d %b %Y')

while start_date < datetime.datetime.now():
    end_date = start_date + relativedelta(years=1, days =-1)
    open_value = df[df['date'] == start_date]['close']
    close_value = df[df['date'] == end_date]['close']
    
    while open_value.empty:
        start_date = start_date + relativedelta(days =+1)
        open_value = df[df['date'] == start_date]['close']

    while close_value.empty:
        end_date = end_date + relativedelta(days =-1)
        close_value = df[df['date'] == end_date]['close']
        
    print(start_date.strftime('%d %b %Y'), '-', end_date.strftime('%d %b %Y'), ' '*5, ((close_value.iloc[0] - open_value.iloc[0]) / open_value.iloc[0] * 100).round(2), '%')
    
    start_date = (start_date + relativedelta(years =+1)).replace(month=4, day=1) 

01 Apr 2005 - 31 Mar 2006       32.02 %
03 Apr 2006 - 30 Mar 2007       26.5 %
02 Apr 2007 - 31 Mar 2008       -31.85 %
01 Apr 2008 - 31 Mar 2009       -35.3 %
01 Apr 2009 - 31 Mar 2010       186.73 %
01 Apr 2010 - 31 Mar 2011       46.56 %
01 Apr 2011 - 30 Mar 2012       -0.96 %
02 Apr 2012 - 28 Mar 2013       31.94 %
01 Apr 2013 - 31 Mar 2014       37.02 %
01 Apr 2014 - 31 Mar 2015       17.33 %
01 Apr 2015 - 31 Mar 2016       -0.88 %
01 Apr 2016 - 31 Mar 2017       -0.9 %
03 Apr 2017 - 28 Mar 2018       18.14 %
02 Apr 2018 - 29 Mar 2019       37.59 %
01 Apr 2019 - 03 May 2019       4.94 %


can say, financial year 2009-10 was the best time for TCS, in which shareholders incresed their invested value by nearly three times and worst hit in the financial years 2007-08 and 2008-09

In [41]:
df.groupby(pd.DatetimeIndex(df['date']).year)['volume'].mean().sort_values(ascending=False).astype(int)

date
2004    11206440
2009     6363109
2005     5968632
2008     5274727
2007     4284960
2006     3982452
2018     3452863
2010     3336359
2011     3313313
2019     3126160
2013     2700566
2012     2664643
2014     2374837
2015     2344592
2017     2210291
2016     2183429
Name: volume, dtype: int32

on an average, nearly 25 to 35 lakhs TCS shares will traded in a single day

In [42]:
df[df['open'] == df['close']]

date     open     high      low    close   volume bonus dividend  \
566  2006-11-24   287.50   289.23   284.04   287.50  3205304   NaN      NaN   
981  2008-07-25   199.80   212.38   195.50   199.80  8842536   NaN      NaN   
1000 2008-08-22   204.75   206.19   200.88   204.75  2645284   NaN      NaN   
1204 2009-07-02   195.00   197.65   190.60   195.00  3715428   NaN      NaN   
1922 2012-05-17   611.00   617.50   599.53   611.00  4137524   NaN      NaN   
2188 2013-06-10   762.53   772.35   759.55   762.53  3360430   NaN      NaN   
3000 2016-09-26  1200.55  1209.55  1197.50  1200.55  1753476   NaN      NaN   

     rights split quarter  
566     NaN   NaN      Q3  
981     NaN   NaN      Q2  
1000    NaN   NaN      Q2  
1204    NaN   NaN      Q2  
1922    NaN   NaN      Q1  
2188    NaN   NaN      Q1  
3000    NaN   NaN      Q2

In [43]:
df[df['high'] == df['close']]

Empty DataFrame
Columns: [date, open, high, low, close, volume, bonus, dividend, rights, split, quarter]
Index: []

In [44]:
df[df['low'] == df['close']]

Empty DataFrame
Columns: [date, open, high, low, close, volume, bonus, dividend, rights, split, quarter]
Index: []

In [60]:
cls_10yrs_ago = df[df['date'] == df['date'].tail(1).iloc[0] - pd.DateOffset(years=10, days=6)]['close'].iloc[0]

In [61]:
(df['close'].tail(1).iloc[0] - cls_10yrs_ago) / cls_10yrs_ago

13.167995746943117

Today's stock value 14 times higher than that of the value 10 yrs ago

In [62]:
div_list = df[df['dividend'].notnull()]['dividend'].values.tolist()

In [63]:
max([int(i.split()[1].rstrip('%')) for i in div_list])

4000

In [64]:
[i for i in div_list if '4000' in i]    

['Special 4000%']

In [65]:
df[df['dividend'] == 'Special 4000%']

date     open     high     low    close   volume bonus  \
2470 2014-07-28  1288.58  1299.83  1277.5  1294.65  1314922   NaN   

           dividend rights split quarter  
2470  Special 4000%    NaN   NaN      Q2

on the completion of 10 yrs of being active in NSE, TCS has given the highest dividend til date which is 4000%

In [66]:
df[df['bonus'].notnull()]

date     open     high     low    close   volume      bonus  \
484  2006-07-28   235.03   243.74   231.0   232.12  3019608  Ratio 1:1   
1192 2009-06-16   191.65   196.75   185.6   194.58  4344006  Ratio 1:1   
3417 2018-05-31  1734.00  1759.05  1726.1  1741.05  5049371  Ratio 1:1   

          dividend rights split quarter  
484   Interim 300%    NaN   NaN      Q2  
1192    Final 500%    NaN   NaN      Q1  
3417   Final 2900%    NaN   NaN      Q1

TCS has given the bonus shares twice in first quarter and once in second quarter and never made an attempt to split the share

In [67]:
df[df['date'] == df['date'].tail(1).iloc[0] - pd.DateOffset(years=10, days=-1)]['close']

1161    166.6
Name: close, dtype: float64

# Analysis of Banking Sector(Private)

In [3]:
os.chdir(r'G:\project\data_03052019\Banks - Private Sector')

d = {}
for i, j in enumerate(os.listdir('.')):
    df_name = j.rstrip('.csv').replace(' ','')
    d[df_name] = pd.read_csv(j, header = None)
    d[df_name]['Stock'] = j.rstrip('.csv')
    d[df_name].columns = ['date', 'open', 'high', 'low', 'close', 'volume', 'Bonus', 'dividend', 'Rights : Ratio', 'split', 'Stock' ]

In [4]:
d.keys()

dict_keys(['AxisBank', 'BandhanBank', 'CityUnionBank', 'DCBBank', 'DhanlaxmiBank', 'FederalBank', 'HDFCBank', 'ICICIBank', 'IDFCFirstBank', 'IndusIndBank', 'JKBank', 'KarnatakaBank', 'KarurVysya', 'KotakMahindra', 'LakshmiVila', 'RETFGoldBeES', 'RBLBank', 'SouthIndBk', 'StanChartPLC', 'YesBank'])

In [5]:
d['AxisBank'].head(3)

date  open  high   low  close   volume  Bonus dividend  \
0  03 Jan 2000  5.34  5.34  5.34   5.34   560500    NaN      NaN   
1  04 Jan 2000  5.40  5.74  5.30   5.37  1172500    NaN      NaN   
2  05 Jan 2000  5.20  5.55  5.10   5.26   850500    NaN      NaN   

   Rights : Ratio split      Stock  
0             NaN   NaN  Axis Bank  
1             NaN   NaN  Axis Bank  
2             NaN   NaN  Axis Bank

In [6]:
d['AxisBank'].sample(3)

date   open    high     low   close    volume  Bonus dividend  \
1950  15 Oct 2007  150.6  166.20  150.40  164.00   6582640    NaN      NaN   
3180  28 Sep 2012  228.4  230.18  226.13  227.20   8744000    NaN      NaN   
2192  03 Oct 2008  145.0  148.00  138.82  140.46  23026825    NaN      NaN   

      Rights : Ratio split      Stock  
1950             NaN   NaN  Axis Bank  
3180             NaN   NaN  Axis Bank  
2192             NaN   NaN  Axis Bank

In [7]:
for i in d.keys():
    d[i]['date'] = pd.to_datetime(d[i]['date'])

In [9]:
result_list = []

for i in d.keys():
    result_list.append((  (d[i]['date'].head(1).iloc[0], d[i]['Stock'].head(1).iloc[0])))
    
for i, j in sorted(result_list):
    print(j , '===' , i.strftime('%d-%b-%Y'))

Axis Bank === 03-Jan-2000
City Union Bank === 03-Jan-2000
Federal Bank === 03-Jan-2000
HDFC Bank === 03-Jan-2000
ICICI Bank === 03-Jan-2000
IndusInd Bank === 03-Jan-2000
JK Bank === 03-Jan-2000
Karur Vysya === 03-Jan-2000
Kotak Mahindra === 03-Jan-2000
Lakshmi Vila === 03-Jan-2000
South Ind Bk === 03-Jan-2000
Karnataka Bank === 23-Jun-2000
Dhanlaxmi Bank === 03-May-2004
Yes Bank === 12-Jul-2005
DCB Bank === 27-Oct-2006
R ETF Gold BeES === 21-Mar-2007
StanChart PLC === 11-Jun-2010
IDFC First Bank === 06-Nov-2015
RBL Bank === 31-Aug-2016
Bandhan Bank === 27-Mar-2018


Could say, in the last 10 years 4 private banks offered IPO and around 10 banks in the private sector running from atleast last 20 years

In [19]:
AllTimeLow_dict = {}

AllTimeLow_dict['stock'] = []
AllTimeLow_dict['date'] = []
AllTimeLow_dict['low_value'] = []

for i in d.keys():
    
    index_pos = d[i]['low'].idxmin()
    AllTimeLow_dict['stock'].append( d[i].iloc[index_pos]['Stock'])
    AllTimeLow_dict['date'].append( d[i].iloc[index_pos]['date'])
    AllTimeLow_dict['low_value'].append( d[i].iloc[index_pos]['low'])

In [12]:
AllTimeLow = pd.DataFrame.from_dict(AllTimeLow_dict)

In [20]:
AllTimeHigh_dict = {}

AllTimeHigh_dict['stock'] = []
AllTimeHigh_dict['date'] = []
AllTimeHigh_dict['high_value'] = []

for i in d.keys():
    
    index_pos = d[i]['high'].idxmax()
    AllTimeHigh_dict['stock'].append( d[i].iloc[index_pos]['Stock'])
    AllTimeHigh_dict['date'].append( d[i].iloc[index_pos]['date'])
    AllTimeHigh_dict['high_value'].append( d[i].iloc[index_pos]['high'])

In [24]:
AllTimehigh = pd.DataFrame.from_dict(AllTimeHigh_dict)

In [25]:
merged_df = pd.merge(AllTimeLow, AllTimehigh, on='stock', how='inner')

In [31]:
merged_df.rename({'date_x' : 'AllTimeLow', 'date_y' : 'AllTimeHigh'}, axis=1, inplace=True)

In [32]:
merged_df

stock AllTimeLow  low_value AllTimeHigh  high_value
0         Axis Bank 2001-09-18       4.20  2019-03-28      787.65
1      Bandhan Bank 2018-10-26     368.10  2018-08-09      741.80
2   City Union Bank 2000-10-11       1.05  2018-05-31      223.69
3          DCB Bank 2009-03-09      13.65  2019-04-25      219.75
4    Dhanlaxmi Bank 2018-10-01      10.70  2010-10-28      212.20
5      Federal Bank 2001-09-17       0.83  2017-10-17      127.65
6         HDFC Bank 2000-01-11      31.40  2019-05-03     2378.00
7        ICICI Bank 2000-01-07      12.00  2019-04-18      411.00
8   IDFC First Bank 2018-10-08      32.70  2016-09-28       83.40
9     IndusInd Bank 2001-09-18       8.50  2018-08-03     2038.00
10          JK Bank 2000-08-18       2.66  2014-05-26      195.48
11   Karnataka Bank 2000-11-01      12.97  2008-02-05      213.56
12      Karur Vysya 2000-01-11       1.60  2017-09-18      136.71
13   Kotak Mahindra 2001-09-18       1.30  2019-05-03     1435.75
14     Lakshmi Vila 2000-03-14       9.28  2017-07-06      194.89
15  R ETF Gold BeES 2007-06-18     797.00  2012-09-14     3119.00
16         RBL Bank 2016-08-31     274.20  2019-03-29      691.45
17     South Ind Bk 2000-05-23       1.03  2018-01-09       34.75
18    StanChart PLC 2016-02-12      34.55  2013-10-11      135.00
19         Yes Bank 2009-03-09       8.12  2018-08-20      404.00

In [34]:
merged_df['max_change'] = (merged_df['high_value'] - merged_df['low_value']) / merged_df['low_value'] * 100

In [40]:
merged_df['duration'] = (merged_df['AllTimeHigh'].dt.year - merged_df['AllTimeLow'].dt.year).apply(str) + ' Yrs ' + ((merged_df['AllTimeHigh'].dt.month - merged_df['AllTimeLow'].dt.month)% 12).apply(str) + ' Months'

In [42]:
merged_df.sort_values('max_change', ascending=False)
#duration is the time taken to achieve it's all time high from all time low

stock AllTimeLow  low_value AllTimeHigh  high_value  \
13   Kotak Mahindra 2001-09-18       1.30  2019-05-03     1435.75   
9     IndusInd Bank 2001-09-18       8.50  2018-08-03     2038.00   
2   City Union Bank 2000-10-11       1.05  2018-05-31      223.69   
0         Axis Bank 2001-09-18       4.20  2019-03-28      787.65   
5      Federal Bank 2001-09-17       0.83  2017-10-17      127.65   
12      Karur Vysya 2000-01-11       1.60  2017-09-18      136.71   
6         HDFC Bank 2000-01-11      31.40  2019-05-03     2378.00   
10          JK Bank 2000-08-18       2.66  2014-05-26      195.48   
19         Yes Bank 2009-03-09       8.12  2018-08-20      404.00   
7        ICICI Bank 2000-01-07      12.00  2019-04-18      411.00   
17     South Ind Bk 2000-05-23       1.03  2018-01-09       34.75   
14     Lakshmi Vila 2000-03-14       9.28  2017-07-06      194.89   
4    Dhanlaxmi Bank 2018-10-01      10.70  2010-10-28      212.20   
11   Karnataka Bank 2000-11-01      12.97  2008-02-05      213.56   
3          DCB Bank 2009-03-09      13.65  2019-04-25      219.75   
15  R ETF Gold BeES 2007-06-18     797.00  2012-09-14     3119.00   
18    StanChart PLC 2016-02-12      34.55  2013-10-11      135.00   
8   IDFC First Bank 2018-10-08      32.70  2016-09-28       83.40   
16         RBL Bank 2016-08-31     274.20  2019-03-29      691.45   
1      Bandhan Bank 2018-10-26     368.10  2018-08-09      741.80   

       max_change          duration  
13  110342.307692   18 Yrs 8 Months  
9    23876.470588  17 Yrs 11 Months  
2    21203.809524   18 Yrs 7 Months  
0    18653.571429   18 Yrs 6 Months  
5    15279.518072   16 Yrs 1 Months  
12    8444.375000   17 Yrs 8 Months  
6     7473.248408   19 Yrs 4 Months  
10    7248.872180   14 Yrs 9 Months  
19    4875.369458    9 Yrs 5 Months  
7     3325.000000   19 Yrs 3 Months  
17    3273.786408   18 Yrs 8 Months  
14    2000.107759   17 Yrs 4 Months  
4     1883.177570   -8 Yrs 0 Months  
11    1546.569005    8 Yrs 3 Months  
3     1509.890110   10 Yrs 1 Months  
15     291.342535    5 Yrs 3 Months  
18     290.738061   -3 Yrs 8 Months  
8      155.045872  -2 Yrs 11 Months  
16     152.169949    3 Yrs 7 Months  
1      101.521326   0 Yrs 10 Months

clearly says, that Kotak Mahindra Bank has seen the maximum up's and downs, it had all time high just a few days ago which is nearly 1100 times to it's all time low followed by IndusInd Bank, both share the common date for their all time low

There are three banks for which the duration is low, could say the are heading in the negative way


In [43]:
result_dict = {}

from dateutil.relativedelta import relativedelta

no_of_days = 365

for i in d.keys():
    start_date = datetime.datetime.strptime('01 Apr 2005', '%d %b %Y')
    
    if not (d[i].tail(1)['date'].iloc[0] - d[i].head(1)['date'].iloc[0]).days < (datetime.datetime.now() - start_date).days:
        return_list = []
        
        while start_date < datetime.datetime.strptime('01 Apr 2019', '%d %b %Y'):
            end_date = start_date + relativedelta(years=1, days =-1)
            open_value = d[i][d[i]['date'] == start_date]['close']
            close_value = d[i][d[i]['date'] == end_date]['close']            
    
            while open_value.empty:
                start_date = start_date + relativedelta(days =+1)
                open_value = d[i][d[i]['date'] == start_date]['close']

            while close_value.empty:
                end_date = end_date + relativedelta(days =-1)
                close_value = d[i][d[i]['date'] == end_date]['close']
                
            return_list.append(((close_value.iloc[0] - open_value.iloc[0]) / open_value.iloc[0] * 100).round(2))
            
    
            start_date = (start_date + relativedelta(years =+1)).replace(month=4, day=1)
        
        else:
            result_dict[i] = return_list

In [44]:
result_df = pd.DataFrame(result_dict, index=range(2006, 2020))

In [47]:
result_df
#all values are in percentage

AxisBank  CityUnionBank  DhanlaxmiBank  FederalBank  HDFCBank  \
2006     49.51          29.25           7.01        30.46     40.38   
2007     35.83          40.05          84.05         6.38     23.30   
2008     71.97          82.50          56.17        39.92     47.70   
2009    -42.50         -56.89         -19.95       -35.29    -25.67   
2010    179.39         171.90         158.20        86.90     93.35   
2011     19.62          54.98         -15.52        57.96     20.97   
2012    -18.58           4.85         -40.89         1.48     11.38   
2013     12.63          23.04         -34.92        11.48     18.36   
2014     11.07           1.22         -26.13        -1.18     20.03   
2015     90.74          83.16         -13.08        40.79     38.56   
2016    -22.03          -3.56         -38.27       -30.26      3.67   
2017      9.09          55.88          48.38        95.61     35.52   
2018      1.98          25.10         -32.92        -1.44     31.60   
2019     55.75           6.54         -21.72         6.93     20.08   

      ICICIBank  IndusIndBank  JKBank  KarnatakaBank  KarurVysya  \
2006      45.06        -10.94   21.05          41.00       18.77   
2007      41.28        -12.23   42.43          67.71       41.67   
2008      -4.30        105.35    8.98          20.60       40.09   
2009     -56.08        -58.82  -53.57         -67.00      -39.79   
2010     172.64        411.58  107.71          76.63      128.13   
2011      17.17         48.76   30.47          -7.05       50.99   
2012     -19.29         21.13    5.23         -12.65       -7.64   
2013      17.38         24.06   31.68          35.36       19.76   
2014      18.40         20.72   26.41         -12.50      -17.39   
2015      28.88         77.47  -41.45           7.18       47.83   
2016     -26.77          4.58  -38.52         -19.68      -19.86   
2017      16.17         48.25   23.97          65.06       26.68   
2018       6.79         28.02  -19.37         -18.54       -7.42   
2019      52.95         -1.36   -9.60          17.13      -23.14   

      KotakMahindra  LakshmiVila  SouthIndBk  
2006         105.78       -43.34       -6.54  
2007          68.35        54.26       58.19  
2008          36.45        31.37       51.72  
2009         -56.07       -35.16      -54.07  
2010         149.82        69.05      238.66  
2011          22.61        20.83       29.59  
2012          20.92       -15.16        6.23  
2013          17.45        -4.25       -0.81  
2014          19.79       -15.40       -9.37  
2015          71.96        95.10       12.26  
2016           0.02       -21.17      -31.19  
2017          28.28        99.65       29.63  
2018          19.62       -37.05        5.80  
2019          21.70       -25.93      -29.64

2009-10 was the best time for banking sectors, in which value of the all stocks almost doubled, where as IndusInd performance was excellent

The financial year 2008-09 banking sector(private) had given the complete negative returns regardless of the stock 

Financial Year 2018-19 was mainly beneficial for Axis and ICICI banks

In [49]:
vol_dict = {}

for i in d.keys():
    vol_dict[i] =  d[i]['volume'].mean()

vol_df = pd.DataFrame(list(vol_dict.items()),  columns = ['stock', 'mean_vol'], dtype= np.int64)

In [52]:
vol_df.sort_values('mean_vol', ascending=False)

stock  mean_vol
7       ICICIBank  16416640
19        YesBank  15331405
8   IDFCFirstBank   9898443
5     FederalBank   6603569
0        AxisBank   6461517
17     SouthIndBk   4407454
6        HDFCBank   2505901
3         DCBBank   2189918
13  KotakMahindra   2146078
11  KarnatakaBank   1744342
1     BandhanBank   1730697
16        RBLBank   1720678
12     KarurVysya   1130269
9    IndusIndBank   1055649
10         JKBank    767869
2   CityUnionBank    620105
4   DhanlaxmiBank    493792
14    LakshmiVila    422428
18   StanChartPLC    212470
15   RETFGoldBeES     33931

On a normal day, we can expect to have the volume traded on ICICI Bank to be around 1,64,00,000 shares and could see least turnover from Reliance ETF Gold BeES

In [59]:
start_date = datetime.datetime.strptime('01 Apr 2018', '%d %b %Y')
end_date = datetime.datetime.strptime('31 Mar 2019', '%d %b %Y')
return_list = []

for i in d.keys():
    temp_df = d[i][(d[i]['date'] >= start_date) & (d[i]['date'] <= end_date)]
    return_list.append((temp_df.tail(1)['close'].iloc[0] - temp_df.head(1)['close'].iloc[0]) / temp_df.head(1)['close'].iloc[0] * 100 )
else:
    print('mean ===', np.array(return_list).mean())
    print('median ===',np.median(np.array(return_list)))

mean === 6.750600796472744
median === 6.733193970175606


Banking sector(private) has given the returns of 6.7% for the financial year 2018-19

In [67]:
current = []
past = []

for i in d.keys():

    for j in range(0, (len(d[i]) - 1)):
        val = d[i]['close'].iloc[ j + 1] - d[i]['close'].iloc[j]
        
        if val > 0:
            current.append(val)
        elif len(current) > len(past):
            past = current
            current = []
            stock = i
        else:
            current = []
else:
    print( stock, len(past))

DCBBank 12


for 12 consecutive days, the stock of DCB Bank has closed positively and this is the positive close for most consecutive days by any stock in banking sector

In [68]:
current = []
past = []

for i in d.keys():

    for j in range(0, (len(d[i]) - 1)):
        val = d[i]['close'].iloc[ j + 1] - d[i]['close'].iloc[j]
        
        if val < 0:
            current.append(val)
        elif len(current) > len(past):
            past = current
            current = []
            stock = i
        else:
            current = []
else:
    print( stock, len(past))

YesBank 14


for 14 consecutive days, the stock of Yes Bank had negative close and this is the negative close for most consecutive days by any stock in banking sector